### ISSUE TITLE
Errors which are caused by heredocs

### ISSUE THREAD

**Issue by jfcherng** [koalaman/shellcheck#647]
```bash
#!/bin/bash

if [ true ]; then cat <<EOF; fi
Condition is met.
EOF
```
```bash
#!/bin/bash

print_info_text_compact () { cat <<EOF; }
This is some info text.
EOF
```
Both of above codes are valid in Bash but shellcheck raises errors which are

![image](https://cloud.githubusercontent.com/assets/6594915/14381561/c8595eae-fdb9-11e5-979a-c4ffc8038a10.png)
![image](https://cloud.githubusercontent.com/assets/6594915/14381556/b7cfa584-fdb9-11e5-9dc6-6acb0c02da04.png)

--

Tested here: http://www.shellcheck.net

**Comment by koalaman** [comment]
Similar to #108, and one of the longest outstanding bugs.

I tried having another stab at it, but quickly found that bash and dash don't do a literal reading of POSIX:

>The here-document shall be treated as a single word that begins after the next <newline> and continues until there is a line containing only the delimiter and a <newline>, with no <blank> characters in between. Then the next here-document starts, if there is one.

That would imply that this is valid:

```
cat - << EOF $(
my heredoc
EOF
)
```

However, bash/dash don't consider the linefeeds inside `$(..)`. Source diving suggests that bash differentiates between command lists with and without significant linefeeds, and processes heredocs after a list without them.

ShellCheck would have to scope out what should be considered the scope for here docs, and then only try to read them after line feeds in that scope.

**Comment by koalaman** [comment]
Fixed in 2f0ae44
